test/unit/app/controllers/permissions/permissions-middleware-test.js

```diff
@@ -0,0 +1,492 @@
+import { strict as assert } from 'assert'
+
+import {
+  METADATA_STORE_KEY,
+} from '../../../../../app/scripts/controllers/permissions/enums'
+
+import {
+  constants,
+  getters,
+  getApprovedPermissionsRequest,
+  getPermController,
+  getPermissionsMiddleware,
+  getUserApprovalPromise,
+  grantPermissions,
+} from './mocks'
+
+const {
+  CAVEATS,
+  ERRORS,
+  PERMS,
+  RPC_REQUESTS,
+} = getters
+
+const {
+  ACCOUNT_ARRAYS,
+  ORIGINS,
+  PERM_NAMES,
+} = constants
+
+const validatePermission = (perm, name, origin, caveats) => {
+  assert.equal(name, perm.parentCapability, 'unexpected permission name')
+  assert.equal(origin, perm.invoker, 'unexpected permission origin')
+  assert.deepEqual(caveats, perm.caveats, 'unexpected permission caveats')
+}
+
+describe('permissions middleware', function () {
+
+  describe('wallet_requestPermissions', function () {
+
+    let permController
+
+    beforeEach(function () {
+      permController = getPermController()
```

Comment:
Hmm - it was throwing me off a bit in the permission controller tests that the module under test wasn't even being imported directly, but I see that it was in the helper function because it was shared between these two tests.

Would it be possible to keep all of the permission controller tests to the other test file, and just test the `methodMiddleware` here? It seems like it'd be simple enough to pass in stubs to the `methodMiddleware` to check each path. That way the helper function for instantiating the permission controller could go directly in with those tests, and it'd be easier to trace from the test directly to the file-under-test. And similarly here as well.

Reply:
The first point is addressed; the `PermissionsController` is now imported in the files that test it.

Regarding testing the `methodMiddleware`: we could certainly test the `methodMiddleware` directly, passing in stubs as necessary, but `permissions-middleware-test` is also testing functionality that has nothing to do with the `methodMiddleware`, for instance `wallet_requestPermissions`.